--- gcc/ada/gcc-interface/Make-lang.in.org	2016-10-11 21:07:08.532970071 +0000
+++ gcc/ada/gcc-interface/Make-lang.in	2016-10-11 21:09:38.974741213 +0000
@@ -995,8 +995,8 @@
 	$(ECHO) "with Osint; use Osint;" >>tmp-sdefault.adb
 	$(ECHO) "package body Sdefault is" >>tmp-sdefault.adb
 	$(ECHO) "   S0 : constant String := \"$(prefix)/\";" >>tmp-sdefault.adb
-	$(ECHO) "   S1 : constant String := \"$(ADA_INCLUDE_DIR)/\";" >>tmp-sdefault.adb
-	$(ECHO) "   S2 : constant String := \"$(ADA_RTL_OBJ_DIR)/\";" >>tmp-sdefault.adb
+	$(ECHO) "   S1 : constant String := \"$(libdir)/gcc/$(target_noncanonical)/4.4/adainclude/\";" >>tmp-sdefault.adb
+	$(ECHO) "   S2 : constant String := \"$(libdir)/gcc/$(target_noncanonical)/4.4/adalib/\";" >>tmp-sdefault.adb
 	$(ECHO) "   S3 : constant String := \"$(target)/\";" >>tmp-sdefault.adb
 	$(ECHO) "   S4 : constant String := \"$(libsubdir)/\";" >>tmp-sdefault.adb
 	$(ECHO) "   function Include_Dir_Default_Name return String_Ptr is" >>tmp-sdefault.adb
