Index: TurboGears-1.1.3/turbogears/identity/base.py
===================================================================
--- TurboGears-1.1.3.orig/turbogears/identity/base.py
+++ TurboGears-1.1.3/turbogears/identity/base.py
@@ -55,10 +55,10 @@ def create_default_provider():
     for entrypoint in plugins:
         try:
             provider_class = entrypoint.load()
-        except Exception:
+        except Exception as load_error:
             raise IdentityConfigurationException(
                 "IdentityProvider plugin can't be loaded: %s\n%s"
-                % (provider_plugin, load_error))
+                % (provider_plugin, repr(load_error)))
         break
     else:
         provider_class = load_class(provider_plugin)
