diff -up TurboGears-1.1.1/turbogears/feed/feed.py.bak TurboGears-1.1.1/turbogears/feed/feed.py
--- TurboGears-1.1.1/turbogears/feed/feed.py.bak	2010-12-02 17:06:50.287780420 -0800
+++ TurboGears-1.1.1/turbogears/feed/feed.py	2010-12-02 12:54:30.501525758 -0800
@@ -16,7 +16,7 @@ class FeedController(turbogears.controll
     """ Object for generating feeds in multiple formats """
 
     def __init__(self, default="atom1.0"):
-        warnings.warn(__doc__, DeprecationWarning)
+        warn(__doc__, DeprecationWarning)
         self.default = default
         self.formats = ["atom1.0", "atom0.3", "rss2.0"]
 
