Index: TurboGears-1.1.3/setup.py
===================================================================
--- TurboGears-1.1.3.orig/setup.py
+++ TurboGears-1.1.3/setup.py
@@ -28,7 +28,7 @@ install_requires = [
     "setuptools >= 0.6c11",
     "simplejson >= 1.9.1",
     "TurboJson >= 1.3",
-    "TurboKid >= 1.0.5", # needed as engine for widget templates
+    "TurboKid >= 1.0.4", # needed as engine for widget templates
     "tgMochiKit >= 1.4.2",
 ]
 
Index: TurboGears-1.1.3/TurboGears.egg-info/requires.txt
===================================================================
--- TurboGears-1.1.3.orig/TurboGears.egg-info/requires.txt
+++ TurboGears-1.1.3/TurboGears.egg-info/requires.txt
@@ -7,7 +7,7 @@ PEAK-Rules >= 0.5a1.dev-r2555
 setuptools >= 0.6c11
 simplejson >= 1.9.1
 TurboJson >= 1.3
-TurboKid >= 1.0.5
+TurboKid >= 1.0.4
 tgMochiKit >= 1.4.2
 
 [compat]
