--- ./gnome-dictionary/libgdict/gdict-utils.c.org	2016-10-20 19:30:01.767222505 +0000
+++ ./gnome-dictionary/libgdict/gdict-utils.c	2016-10-20 19:30:35.567121558 +0000
@@ -279,7 +279,7 @@
   GtkWidget *toplevel;
   
   toplevel = gtk_widget_get_toplevel (widget);
-  if (!GTK_WIDGET_TOPLEVEL (toplevel))
+  if (!gtk_widget_is_toplevel (toplevel))
     return NULL;
   else
     return GTK_WINDOW (toplevel);
