--- java-1.6.0-openjdk.spec.org	2017-03-17 15:55:47.288308381 +0000
+++ java-1.6.0-openjdk.spec	2017-03-17 15:58:57.538992980 +0000
@@ -66,6 +66,10 @@
 %define archbuild sparcv9
 %define archinstall sparcv9
 %endif
+%ifarch %{arm}
+%define archbuild arm
+%define archinstall arm
+%endif
 %ifnarch %{jit_arches}
 %define archbuild %{_arch}
 %define archinstall %{_arch}
@@ -249,7 +255,7 @@
 BuildRequires: libffi-devel
 %endif
 
-ExclusiveArch: x86_64 i686
+ExclusiveArch: x86_64 i686 %{arm}
 
 # cacerts build requirement.
 BuildRequires: openssl
