diff -up ImageMagick-6.7.2-7/magick/transform.c.splice-crash ImageMagick-6.7.2-7/magick/transform.c
--- ImageMagick-6.7.2-7/magick/transform.c.splice-crash	2011-09-16 03:38:06.000000000 +0200
+++ ImageMagick-6.7.2-7/magick/transform.c	2016-06-03 15:25:23.329536796 +0200
@@ -65,6 +65,7 @@
 #include "magick/string_.h"
 #include "magick/thread-private.h"
 #include "magick/transform.h"
+#define MagickMin(x,y) (((x) < (y)) ? (x) : (y))
 
 /*
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
@@ -1635,6 +1636,7 @@ MagickExport Image *SpliceImage(const Im
     splice_geometry;
 
   ssize_t
+    columns,
     y;
 
   /*
@@ -1719,6 +1721,7 @@ MagickExport Image *SpliceImage(const Im
   */
   status=MagickTrue;
   progress=0;
+  columns=MagickMin(splice_geometry.x,(ssize_t) splice_image->columns);
   image_view=AcquireCacheView(image);
   splice_view=AcquireCacheView(splice_image);
 #if defined(MAGICKCORE_OPENMP_SUPPORT)
@@ -1741,7 +1744,8 @@ MagickExport Image *SpliceImage(const Im
 
     if (status == MagickFalse)
       continue;
-    p=GetCacheViewVirtualPixels(image_view,0,y,image->columns,1,exception);
+    p=GetCacheViewVirtualPixels(image_view,0,y,splice_image->columns,1,
+      exception);
     q=QueueCacheViewAuthenticPixels(splice_view,0,y,splice_image->columns,1,
       exception);
     if ((p == (const PixelPacket *) NULL) || (q == (PixelPacket *) NULL))
@@ -1751,7 +1755,7 @@ MagickExport Image *SpliceImage(const Im
       }
     indexes=GetCacheViewAuthenticIndexQueue(image_view);
     splice_indexes=GetCacheViewAuthenticIndexQueue(splice_view);
-    for (x=0; x < splice_geometry.x; x++)
+    for (x=0; x < columns; x++)
     {
       SetPixelRed(q,GetPixelRed(p));
       SetPixelGreen(q,GetPixelGreen(p));
@@ -1820,10 +1824,10 @@ MagickExport Image *SpliceImage(const Im
 
     if (status == MagickFalse)
       continue;
-    p=GetCacheViewVirtualPixels(image_view,0,y-(ssize_t) splice_geometry.height,
-      image->columns,1,exception);
-    if ((y < 0) || (y >= (ssize_t) splice_image->rows))
+    if ((y < 0) || (y >= (ssize_t)splice_image->rows))
       continue;
+    p=GetCacheViewVirtualPixels(image_view,0,y-(ssize_t) splice_geometry.height,
+      splice_image->columns,1,exception);
     q=QueueCacheViewAuthenticPixels(splice_view,0,y,splice_image->columns,1,
       exception);
     if ((p == (const PixelPacket *) NULL) || (q == (PixelPacket *) NULL))
@@ -1833,7 +1837,7 @@ MagickExport Image *SpliceImage(const Im
       }
     indexes=GetCacheViewAuthenticIndexQueue(image_view);
     splice_indexes=GetCacheViewAuthenticIndexQueue(splice_view);
-    for (x=0; x < splice_geometry.x; x++)
+    for (x=0; x < columns; x++)
     {
       SetPixelRed(q,GetPixelRed(p));
       SetPixelGreen(q,GetPixelGreen(p));
