diff -ruN rpmbuild-orig/SOURCES/firefox-redsleeve-default-prefs.js rpmbuild/SOURCES/firefox-redsleeve-default-prefs.js
--- rpmbuild-orig/SOURCES/firefox-redsleeve-default-prefs.js	1970-01-01 01:00:00.000000000 +0100
+++ rpmbuild/SOURCES/firefox-redsleeve-default-prefs.js	2015-11-05 20:48:57.246925518 +0100
@@ -0,0 +1,28 @@
+pref("app.update.auto",                     false);
+pref("app.update.enabled",                  false);
+pref("app.update.autoInstallEnabled",       false);
+pref("browser.backspace_action",            2);
+pref("browser.display.use_system_colors",   true);
+pref("browser.download.folderList",         1);
+pref("browser.link.open_external",          3);
+pref("browser.shell.checkDefaultBrowser",   false);
+pref("general.smoothScroll",                true);
+pref("general.useragent.vendor",            "RedSleeve");
+pref("general.useragent.vendorSub",         "FIREFOX_RPM_VR");
+pref("intl.locale.matchOS",                 true);
+pref("storage.nfs_filesystem",              false);
+pref("dom.ipc.plugins.enabled.nswrapper*",  false);
+pref("network.manage-offline-status",       true);
+pref("toolkit.networkmanager.disable", false);
+pref("browser.startup.homepage",            "data:text/plain,browser.startup.homepage=file:///usr/share/doc/HTML/index.html");
+pref("toolkit.storage.synchronous",         0);
+pref("startup.homepage_override_url",       "http://www.redsleeve.org");
+pref("startup.homepage_welcome_url",        "http://www.redsleeve.org");
+/* Workaround for rhbz#1134876 */
+pref("javascript.options.baselinejit",      false);
+pref("extensions.shownSelectionUI",         true);
+/* Workaround for rhbz#1110291 */
+pref("network.negotiate-auth.allow-insecure-ntlm-v1", true);
+/* Workaround for mozbz#1063315 */
+pref("security.use_mozillapkix_verification", false);
+pref("geo.wifi.uri", "https://location.services.mozilla.com/v1/geolocate?key=%MOZILLA_API_KEY%");
diff -ruN rpmbuild-orig/SPECS/firefox.spec rpmbuild/SPECS/firefox.spec
--- rpmbuild-orig/SPECS/firefox.spec	2015-11-04 13:21:45.000000000 +0100
+++ rpmbuild/SPECS/firefox.spec	2015-11-05 20:51:55.756471483 +0100
@@ -54,7 +54,7 @@
 Summary:        Mozilla Firefox Web browser
 Name:           firefox
 Version:        38.4.0
-Release:        1%{?prever}%{?dist}
+Release:        1%{?prever}%{?dist}.redsleeve
 URL:            http://www.mozilla.org/projects/firefox/
 License:        MPLv1.1 or GPLv2+ or LGPLv2+
 Group:          Applications/Internet
@@ -65,7 +65,7 @@
 %endif
 Source10:       firefox-mozconfig
 Source11:       firefox-mozconfig-branded
-Source12:       firefox-centos-default-prefs.js
+Source12:       firefox-redsleeve-default-prefs.js
 Source20:       firefox.desktop
 Source21:       firefox.sh.in
 Source23:       firefox.1
@@ -533,6 +533,9 @@
 #---------------------------------------------------------------------
 
 %changelog
+* Thu Nov 05 2015 Jacco Ligthart <jacco@redsleeve.org> - 38.4.0-1.el7.redsleeve
+- redsleeve default prefs
+
 * Wed Nov 04 2015 CentOS Sources <bugs@centos.org> - 38.4.0-1.el7.centos
 - CentOS default prefs
 
