diff -urNp orig-net-snmp-5.5/snmplib/snmp_api.c new-net-snmp-5.5/snmplib/snmp_api.c
--- orig-net-snmp-5.5/snmplib/snmp_api.c	2016-08-01 13:10:31.896946464 +0200
+++ new-net-snmp-5.5/snmplib/snmp_api.c	2016-08-01 13:17:38.078766623 +0200
@@ -5107,7 +5107,7 @@ _sess_async_send(void *sessp,
      * specified in the received PDU.  
      */
 
-    if (session->sndMsgMaxSize != 0 && length > session->sndMsgMaxSize) {
+    if (pdu->version == SNMP_VERSION_3 && session->sndMsgMaxSize != 0 && length > session->sndMsgMaxSize) {
         DEBUGMSGTL(("sess_async_send",
                     "length of packet (%lu) exceeds session maximum (%lu)\n",
                     (unsigned long)length, (unsigned long)session->sndMsgMaxSize));
