--- acpid.spec.org	2016-08-29 17:01:37.886839281 +0000
+++ acpid.spec	2016-08-29 17:02:31.091473076 +0000
@@ -12,7 +12,7 @@
 Source5: acpid.sysconfig
 Patch1: acpid-1.0.8-makefile.patch
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
-ExclusiveArch: ia64 x86_64 %{ix86}
+ExclusiveArch: ia64 x86_64 %{ix86} %{arm}
 URL: http://acpid.sourceforge.net/
 Requires(post): /sbin/chkconfig
 Requires(preun): /sbin/chkconfig
